### DISCUSSION TITLE
Remote control of Macos lag

### DISCUSSION THREAD

**Discussion by dreamfly-jiefa** [rustdesk/rustdesk#10050] [General]
### Bug Description

1. The controlled end is MacBook Air M1, MacOS 12.7
2. The control end is Windows, Android, iOS
3. Controlling Windows is very smooth in the same network environment, but controlling macOS may cause lag
4. I have tried switching the resolution and encoding method on the Macos side, but it has not been resolved. The maximum frame rate on Macos is up to 10 frames per second.
5. For specific phenomena, please refer to the video link below.
https://www.youtube.com/watch?v=yXrdYJaukSM

### How to Reproduce

1. The controlled end is MacBook Air M1, MacOS 12.7
2. The control end is Windows, Android, iOS
3. Controlling Windows is very smooth in the same network environment, but controlling macOS may cause lag
4. I have tried switching the resolution and encoding method on the Macos side, but it has not been resolved. The maximum frame rate on Macos is up to 10 frames per second.
5. For specific phenomena, please refer to the video link below.
https://www.youtube.com/watch?v=yXrdYJaukSM

### Expected Behavior

Controlled end macOS needs to run smoothly

### Operating system(s) on local (controlling) side and remote (controlled) side

Windows 10->MacOS 12.7

### RustDesk Version(s) on local (controlling) side and remote (controlled) side

Any->1.3.2

**Comment by 21pages** [comment]
What about different codecs? And could you show the controlled side log? Remember to remove private informations. https://github.com/rustdesk/rustdesk/wiki/FAQ#access-logs

**Comment by dreamfly-jiefa** [comment]
> What about different codecs? And could you show the controlled side log? Remember to remove private informations. https://github.com/rustdesk/rustdesk/wiki/FAQ#access-logs

If it is set to AV1 encoding mode, the frame rate can reach 20 frames, but the picture is also too laggy.
Here are the log files (for macOS)
[RustDesk_rCURRENT.log](https://github.com/user-attachments/files/17900599/RustDesk_rCURRENT.log)

**Reply by 21pages** [reply]
Reproduced, encoding is slow with retina enabled.

**Reply by dreamfly-jiefa** [reply]
May I ask how to solve this?

**Reply by 21pages** [reply]
> Reproduced, encoding is slow with retina enabled.

Sorry, I run it in debug mode, in release mode, it's not slow.

**Reply by dreamfly-jiefa** [reply]
How do I enable debug mode or release  mode？Thank you

**Reply by dreamfly-jiefa** [reply]
Today I updated my MacBook Air M1 to the latest system macOS 15.1.1. Installed the latest client version 1.3.3 of Rustdesk. Discovered the following issue: 
1. Remote control of MacBook using iOS client, the frame rate can reach 60 frames per second or even higher. 
2. Besides the iOS client, using Android and Windows clients to control MacBook also results in a frame rate of less than 20 frames per second. 
3. May I ask how to solve this?
@21pages

**Comment by mahergreg** [comment]
I just spent AGES trying to fix my lag issues in Rustdesk (Windows 11 -> MacOS m4 Pro).
I finally managed to get it to a stable 30-35 FPS by following the steps below.  

**Note:** running at the native 4k is best, but personally I find the display too small to work with.  **Most of the lag problems I encountered seem to be stemming from Windows OS scaling**.  Its important to match the client monitor (Windows) resolution to that of the host (MacOS).

I hope this saves others some time.

### My Setup:
- RustDesk 1.3.9
- 2.5Ghz network switch
- Self hosted via QNAP nas drive (Docker)
- 1x 4K UHD monitor 
- Host: 
  - Mac Mini M4 Pro
  - 1 x 4K [HDMI Dummy Plug](https://www.amazon.com.au/Duttek-Emulator-Headless-Perforated-Supports/dp/B0D9W2HHM1/ref=pd_lpo_d_sccl_1/358-7711374-7744534?pd_rd_w=DeswB&content-id=amzn1.sym.7ef4c059-30c9-4984-b131-e38134d0b653&pf_rd_p=7ef4c059-30c9-4984-b131-e38134d0b653&pf_rd_r=49HMNDYGW5AWZTZD7KW2&pd_rd_wg=x1nuk&pd_rd_r=42433011-70b6-4950-9d49-7240453119d9&pd_rd_i=B0D9W2HHM1&psc=1https://www.amazon.com.au/Duttek-Emulator-Headless-Perforated-Supports/dp/B0D9W2HHM1/ref=pd_lpo_d_sccl_1/358-7711374-7744534?pd_rd_w=DeswB&content-id=amzn1.sym.7ef4c059-30c9-4984-b131-e38134d0b653&pf_rd_p=7ef4c059-30c9-4984-b131-e38134d0b653&pf_rd_r=49HMNDYGW5AWZTZD7KW2&pd_rd_wg=x1nuk&pd_rd_r=42433011-70b6-4950-9d49-7240453119d9&pd_rd_i=B0D9W2HHM1&psc=1) (possibly required if no monitor connected) . 
  - I got x2 from [AliExpress for $5 NZD](https://www.amazon.com.au/Duttek-Emulator-Headless-Perforated-Supports/dp/B0D9W2HHM1/ref=pd_lpo_d_sccl_1/358-7711374-7744534?pd_rd_w=DeswB&content-id=amzn1.sym.7ef4c059-30c9-4984-b131-e38134d0b653&pf_rd_p=7ef4c059-30c9-4984-b131-e38134d0b653&pf_rd_r=49HMNDYGW5AWZTZD7KW2&pd_rd_wg=x1nuk&pd_rd_r=42433011-70b6-4950-9d49-7240453119d9&pd_rd_i=B0D9W2HHM1&psc=1)
- Client: 
  - Surface Laptop Studio (Windows 11 Pro)

### Steps to Fix lag:
- On Mac
  - Install [BetterDisplay ](https://github.com/waydabber/BetterDisplay )on the Mac
  - Set  resolution to 2560x1440
  - Disable Hi Resolution
  ![image](https://github.com/user-attachments/assets/603257a7-e07f-4394-bd94-0b40bee300ea)

- On Windows
  - Set Monitor Resolution to match the MacOS (2560x1440)
  - Open properties of RustDesk, edit compatibility settings & allow the app to control scaling
  - ![image](https://github.com/user-attachments/assets/912a24f3-862d-4770-88f7-a09564dca0e2)
  - Start Rustdesk & set scaling to "Scaling Original & Image Quality to "Custom)
  ![image](https://github.com/user-attachments/assets/c7af4761-6e0d-493c-96a4-932deefa85f3)
  - Set FPS to 60
  ![image](https://github.com/user-attachments/assets/d4c3f44a-fa48-42a0-9024-cd4437818cc3)

### Result:
![image](https://github.com/user-attachments/assets/4b7c75b8-38fe-43e1-abb3-a66452b9f236)

**Reply by mahergreg** [reply]
Update:
Setting FPS to 120FPS (in the custom options), I can now achieve ~55-60 FPS
![image](https://github.com/user-attachments/assets/0ad88b51-5e3f-4dc5-a87f-487561052087)
